pkg/node/announce.go

```diff
@@ -50,43 +50,94 @@ func AnnounceWithNamePrefix(prefix string, a Announcer) Announcer {
 
 // AnnounceContext returns a new Announcer that undoes any announcements when ctx is Done.
 // This leaks a go routine if ctx is never done.
+//
+// Deprecated: AnnounceContext does not allow synchronising with the undos, which is usually necessary.
+// Use AnnounceScope instead, or a ReplaceAnnouncer if multiple generations are required.
 func AnnounceContext(ctx context.Context, a Announcer) Announcer {
-	mu := sync.Mutex{}
-	var undos []Undo
-	undone := make(chan struct{})
-	go func() {
-		defer close(undone)
-		<-ctx.Done()
-		mu.Lock()
-		defer mu.Unlock()
-		for _, undo := range undos {
-			undo()
-		}
-		undos = nil
-	}()
-	return AnnouncerFunc(func(name string, features ...Feature) Undo {
-		select {
-		case <-ctx.Done():
-			return NilUndo
-		default:
-		}
+	scoped, undo := AnnounceScope(a)
+	_ = context.AfterFunc(ctx, undo)
+	return scoped
+}
+
+// AnnounceScope returns a new scoped Announcer that will undo all announcements made to it when the returned Undo is called.
+// The Undo will block until all undos are complete.
+// Once the Undo is called, the Announcer is no longer valid and any further calls to Announce will do nothing.
+//
+// It is safe to call the Undo more than once - calls after the first will do nothing.
```

Comment:
Worth mentioning that any Undos returned by Announce calls can be Undone early and safely if desired?